### ISSUE TITLE
Behavior/documentation of ranges of length zero is confusing.

### ISSUE THREAD

**Issue by rednum** [rust-lang/rust#37568]
Hi, 

recently in a project I'm contributing to topic of zero length ranges came up:
https://github.com/rust-bio/rust-bio/pull/95
It seems to me like the documentation for this case is confusing. On one hand, docs for `Range` state that:
```
A (half-open) range which is bounded at both ends: { x | start <= x < end }.
```
When `start = end`, we get `{x | start <= x < start}`, ie. an empty set. So according to this definition, it sounds reasonable to me that all ranges of form `(x..x)` are in fact the same object (empty set). But we actually then just get `Range { start: x, end: x }` object. As this struct is table, I suppose this is intended and not going to change (right?), but I don't find this quite obvious. Similar concerns apply when `start > end` (in that case I'd expect panic, even though the doc says nothing about `..` panicing ever). I suggest explicitly mentioning in documentation what happens when `start >= end`.

**Comment by bluss** [comment]
In the standard library, the range types are completely generic, and they can contain any values. Thus your own library may invent a meaning for a range where start > end, or for a range where each endpoint is a hash table, for that matter. It is thus not possible or proper for range syntax itself to panic.

Now onto index ranges as they are used in libstd:

An index range is more than a set of elements it includes. For an example of what I mean by that:

```rust
let data = [1, 2, 3];
let a = &data[1..1];
let b = &data[10..10];
```

The slicing `a` is in bounds and produces an empty slice that points to the element at `1`; the slicing `b` is out of bounds and panics because of this.

The condition for a valid slicing is `start <= end && end <= length`.

**Comment by rednum** [comment]
Thanks for an explanation, I understand this behavior can be useful. However, when you say that "An index range is more than a set of elements it includes.", it is contradictory to what the doc says "A (half-open) range which is bounded at both ends: { x | start <= x < end }.", since "{ x | start <= x < end}", which means "set of elements not less than start and less than end". So the doc says it is a set of things, but as you've pointed out it clearly isn't. So I think set definition should be replaced with something like: "A (half-open) range which contains all elements x such that x >= start and x < end. The ranges where start >= end are valid and can have useful meanings. Use two dots notation, see examples in contains() [and an example similar to the one you've just shown me should be there]".

**Comment by Mark-Simulacrum** [comment]
@rust-lang/docs, do we want to do something about this? It seems to me that the status quo is fairly correct, but there are some interesting points raised here. I don't follow the discussion all that well, though.

**Comment by steveklabnik** [comment]
I'm fine with trying to clarify something.

**Comment by rednum** [comment]
I've reread my [second comment](https://github.com/rust-lang/rust/issues/37568#issuecomment-258368492) and to be honest I'm not 100% sure what point I've tried to make six months ago, sorry for confusion.

My original concern still stands though: I'm confused by using set notation for objects that are not quite sets, especially when it comes to corner cases. In other words: when I see something {x | y < x <= y }, I expect that for each y this is the same object: the empty set (range), so at first glance I expected that both `1..1` and `2..1` would be the same thing. I think that in past I've actually encountered an API that acted just like that, which may contribute to my confusion.

To sum up, I think the doc would be clearer if it explicitly included something about case when start >= end. For example something similar to what @bluss told me: "This object is more than just a set of elements it includes. It's perfectly valid to have ranges where start >= end, see example below...". I can try to phrase it nicely and submit a PR later this week if you think that my reasoning makes any sense.

**Comment by steveklabnik** [comment]
Triage: no changes. 

The crux of it is:

> To sum up, I think the doc would be clearer if it explicitly included something about case when start >= end.

**Comment by steveklabnik** [comment]
Triage: these docs have changed; they no longer use set notation, and specifically says

```text
/// The `Range` `start..end` contains all values with `x >= start` and
/// `x < end`.  It is empty unless `start < end`.
```

As such, I think this is now fixed. Thank you!
